// @react-mapboxgl/core/es/Control.js
import _isEqual from 'lodash/isEqual';
import _omit from 'lodash/omit';

var _class, _temp;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React from 'react';
import PropTypes from 'prop-types';
var Control = (_temp = _class = function (_React$PureComponent) {
  _inherits(Control, _React$PureComponent);

  function Control() {
    _classCallCheck(this, Control);

    return _possibleConstructorReturn(this, _React$PureComponent.apply(this, arguments));
  }

  Control.prototype.createControl = function createControl(props) {
    var mapboxgl = this.context.mapboxgl;

    var ControlClass = typeof props.type === 'string' ? mapboxgl[props.type + 'Control'] : props.type;
    var options = _omit(props, 'type', 'position');

    return new ControlClass(options);
  };

  Control.prototype.componentDidMount = function componentDidMount() {
    var map = this.context.map;

    this.control = this.createControl(this.props);
    map.addControl(this.control, this.props.position);
  };

  Control.prototype.componentWillUnmount = function componentWillUnmount() {
    var map = this.context.map;

    map.removeControl(this.control);
  };

  Control.prototype.componentWillReceiveProps = function componentWillReceiveProps(nextProps) {
    var map = this.context.map;

    if (!_isEqual(this.props, nextProps)) {
      map.removeControl(this.control);
      this.control = this.createControl(nextProps);
      map.addControl(this.control, nextProps.position);
    }
  };

  Control.prototype.render = function render() {
    return null;
  };

  return Control;
}(React.PureComponent), _class.propTypes = {
  type: PropTypes.oneOfType([PropTypes.oneOf(['Navigation', 'Geolocate', 'Attribution', 'Scale', 'Fullscreen']), PropTypes.object, PropTypes.func]).isRequired,
  position: PropTypes.string
}, _class.defaultProps = {
  position: 'top-right'
}, _class.contextTypes = {
  map: PropTypes.object,
  mapboxgl: PropTypes.object
}, _temp);


export default Control;
