// @react-mapboxgl/core/es/Popup.js
import _isNil from 'lodash/isNil';
import _pick from 'lodash/pick';
import _omitBy from 'lodash/omitBy';
import _extend from 'lodash/extend';
import _isEqual from 'lodash/isEqual';

var _class, _temp;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';


/**
 * Open a mapbox-gl 'native' popup.
 *
 * Note: It's usually not going to be a good idea to use the `closeButton` or
 * `closeOnClick` options, because they can conflict with the mounted state of
 * your popup. You should instead 'close' your own <Popup> component by simply
 * unrendering it. If you want a close button, it would be better to render and
 * manage it yourself (tied to some state that controls the rendering of the
 * Popup).
 */
var Popup = (_temp = _class = function (_React$Component) {
  _inherits(Popup, _React$Component);

  function Popup() {
    _classCallCheck(this, Popup);

    return _possibleConstructorReturn(this, _React$Component.apply(this, arguments));
  }

  Popup.prototype.shouldComponentUpdate = function shouldComponentUpdate(nextProps, nextState) {
    return !_isEqual(this.props, nextProps) || !_isEqual(this.state, nextState);
  };

  Popup.prototype.componentDidMount = function componentDidMount() {
    // Setup popup div.
    this.el = document.createElement('div');
    this.el.className = 'react-mapbox--popup';
    _extend(this.el.style, {
      position: 'relative',
      display: 'inline-block'
    });

    // Add popup to map.
    this.addPopup(this.props);
  };

  Popup.prototype.componentWillUnmount = function componentWillUnmount() {
    if (this.component) {
      ReactDOM.unmountComponentAtNode(this.el);
      this.component = null;
    }
    this.removePopup();
    this.el = null;
  };

  Popup.prototype.componentWillReceiveProps = function componentWillReceiveProps(nextProps) {
    // For options or content, we need to make a new popup.
    if (!_isEqual(this.getOptions(this.props), this.getOptions(nextProps)) || !_isEqual(this.props.html, nextProps.html) || !_isEqual(this.props.text, nextProps.text) || !_isEqual(this.props.onClose, nextProps.onClose)) {
      this.removePopup();
      this.addPopup(nextProps);
      return;
    }

    // Otherwise update the current popup.
    if (!_isEqual(this.props.coordinates, nextProps.coordinates)) {
      this.popup.setLngLat(nextProps.coordinates);
    }
    if (this.props.children !== nextProps.children) {
      this.renderChildren(nextProps);
    }
  };

  Popup.prototype.getOptions = function getOptions(props) {
    return _omitBy(_pick(props, ['closeButton', 'closeOnClick', 'anchor', 'offset']), _isNil);
  };

  Popup.prototype.addPopup = function addPopup(props) {
    var _context = this.context,
        map = _context.map,
        mapboxgl = _context.mapboxgl;

    this.popup = new mapboxgl.Popup(this.getOptions(props));
    if (this.props.onClose) {
      this.popup.on('close', this.props.onClose);
    }
    if (props.coordinates) {
      this.popup.setLngLat(props.coordinates);
    }
    if (props.html) {
      this.popup.setHTML(props.html);
    }
    if (props.text) {
      this.popup.setText(props.text);
    }
    if (props.children) {
      this.renderChildren(props);
      this.popup.setDOMContent(this.el);
    }
    this.popup.addTo(map);
  };

  Popup.prototype.removePopup = function removePopup() {
    if (this.popup.isOpen()) {
      this.popup.off('close');
      this.popup.remove();
    }
    this.popup = null;
  };

  Popup.prototype.renderChildren = function renderChildren(props) {
    if (this.component) {
      ReactDOM.unmountComponentAtNode(this.el);
    }
    if (props.children) {
      this.component = ReactDOM.unstable_renderSubtreeIntoContainer(this, props.children, this.el);
    }
  };

  Popup.prototype.render = function render() {
    return null;
  };

  return Popup;
}(React.Component), _class.propTypes = {
  // Popup options.
  closeButton: PropTypes.bool,
  closeOnClick: PropTypes.bool,
  anchor: PropTypes.string,
  offset: PropTypes.oneOfType([PropTypes.number, PropTypes.array, PropTypes.object]),

  // Coordinates.
  coordinates: PropTypes.array,

  // Only one of the following.
  text: PropTypes.string,
  html: PropTypes.string,
  children: PropTypes.node,

  // Events.
  onClose: PropTypes.func
}, _class.defaultProps = {
  closeButton: false,
  closeOnClick: false
}, _class.contextTypes = {
  map: PropTypes.object,
  mapboxgl: PropTypes.object
}, _temp);


export default Popup;
