// @react-mapboxgl/core/es/Source.js
import _isNil from 'lodash/isNil';
import _omitBy from 'lodash/omitBy';
import _extend from 'lodash/extend';
import _pick from 'lodash/pick';
import _isEqual from 'lodash/isEqual';

var _class, _temp;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React from 'react';
import PropTypes from 'prop-types';
var Source = (_temp = _class = function (_React$Component) {
  _inherits(Source, _React$Component);

  function Source() {
    _classCallCheck(this, Source);

    return _possibleConstructorReturn(this, _React$Component.apply(this, arguments));
  }

  Source.prototype.shouldComponentUpdate = function shouldComponentUpdate(nextProps, nextState) {
    return !_isEqual(this.props, nextProps) || !_isEqual(this.state, nextState);
  };

  Source.prototype.componentDidMount = function componentDidMount() {
    this.addSource(this.props);
  };

  Source.prototype.componentWillUnmount = function componentWillUnmount() {
    this.removeSource(this.props);
  };

  Source.prototype.componentWillReceiveProps = function componentWillReceiveProps(nextProps) {
    var map = this.context.map;

    // If any of these props change, we have to just recreate the
    // source from scratch.

    var invalidate = ['type', 'url', 'tiles', 'tileSize', 'minzoom', 'maxzoom', 'buffer', 'tolerance', 'cluster', 'clusterRadius', 'clusterMaxZoom', 'canvas', 'animate'];
    if (!_isEqual(_pick(this.props, invalidate), _pick(nextProps, invalidate))) {
      this.removeSource(this.props.id);
      this.addSource(nextProps);
      return;
    }

    // The coordinates changed.
    if (!_isEqual(this.props.coordinates, nextProps.coordinates)) {
      map.getSource(nextProps.id).setCoordinates(nextProps.coordinates);
    }

    // The data changed.
    if (!_isEqual(this.props.data, nextProps.data)) {
      map.getSource(nextProps.id).setData(nextProps.data);
    }

    // The dataRev changed (perhaps a timestamp). Use this to 'refresh'
    // an external geojson source.
    if (this.props.dataRev !== nextProps.dataRev) {
      map.getSource(nextProps.id).setData(nextProps.data);
    }
  };

  Source.prototype.addSource = function addSource(props) {
    var map = this.context.map;

    var options = {};

    // Grab basic options from props.
    _extend(options, _omitBy(_pick(props, ['type', 'data', 'url', 'tiles', 'tileSize', 'minzoom', 'maxzoom', 'coordinates', 'buffer', 'tolerance', 'cluster', 'clusterRadius', 'clusterMaxZoom', 'canvas', 'animate']), _isNil));

    // Add the source.
    map.addSource(props.id, options);
    map.fire('_addSource', this.props.id);
  };

  Source.prototype.removeSource = function removeSource(props) {
    var map = this.context.map;

    map.removeSource(props.id);
    map.fire('_removeSource', props.id);
  };

  Source.prototype.render = function render() {
    return null;
  };

  return Source;
}(React.Component), _class.propTypes = {
  id: PropTypes.string.isRequired,
  type: PropTypes.oneOf(['canvas', 'geojson', 'image', 'raster', 'vector', 'video']).isRequired,
  data: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  dataRev: PropTypes.number,
  url: PropTypes.string,
  tiles: PropTypes.array,
  tileSize: PropTypes.number,
  minzoom: PropTypes.number,
  maxzoom: PropTypes.number,
  coordinates: PropTypes.array,
  buffer: PropTypes.number,
  tolerance: PropTypes.number,
  cluster: PropTypes.bool,
  clusterRadius: PropTypes.number,
  clusterMaxZoom: PropTypes.number,
  canvas: PropTypes.string,
  animate: PropTypes.bool
}, _class.contextTypes = {
  map: PropTypes.object
}, _temp);


export default Source;
