// @react-mapboxgl/core/lib/Image.js
'use strict';

exports.__esModule = true;

var _omit2 = require('lodash/omit');

var _omit3 = _interopRequireDefault(_omit2);

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Image = (_temp = _class = function (_React$PureComponent) {
  _inherits(Image, _React$PureComponent);

  function Image() {
    _classCallCheck(this, Image);

    return _possibleConstructorReturn(this, _React$PureComponent.apply(this, arguments));
  }

  Image.prototype.componentDidMount = function componentDidMount() {
    var map = this.context.map;

    map.addImage(this.props.name, this.props.image, (0, _omit3.default)(this.props, 'name', 'image'));
  };

  Image.prototype.componentWillUnmount = function componentWillUnmount() {
    var map = this.context.map;

    map.removeImage(this.props.name);
  };

  Image.prototype.render = function render() {
    return null;
  };

  return Image;
}(_react2.default.PureComponent), _class.propTypes = {
  name: _react2.default.PropTypes.string,
  image: _react2.default.PropTypes.any // HTMLImageElement | ArrayBufferView
}, _class.contextTypes = {
  map: _propTypes2.default.object
}, _temp);
exports.default = Image;
