// @react-mapboxgl/core/lib/MapEvent.js
'use strict';

exports.__esModule = true;

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var MapEvent = (_temp = _class = function (_React$PureComponent) {
  _inherits(MapEvent, _React$PureComponent);

  function MapEvent() {
    _classCallCheck(this, MapEvent);

    return _possibleConstructorReturn(this, _React$PureComponent.apply(this, arguments));
  }

  MapEvent.prototype.componentDidMount = function componentDidMount() {
    var map = this.context.map;

    map.on(this.props.type, this.props.onChange);
  };

  MapEvent.prototype.componentWillUnmount = function componentWillUnmount() {
    var map = this.context.map;

    map.off(this.props.type, this.props.onChange);
  };

  MapEvent.prototype.componentWillReceiveProps = function componentWillReceiveProps(nextProps) {
    var map = this.context.map;

    if (this.props.type !== nextProps.type || this.props.onChange !== nextProps.onChange) {
      map.off(this.props.type, this.props.onChange);
      map.on(nextProps.type, nextProps.onChange);
    }
  };

  MapEvent.prototype.render = function render() {
    return null;
  };

  return MapEvent;
}(_react2.default.PureComponent), _class.propTypes = {
  type: _propTypes2.default.string.isRequired,
  onChange: _propTypes2.default.func.isRequired
}, _class.contextTypes = {
  map: _propTypes2.default.object
}, _temp);
exports.default = MapEvent;
