// @react-mapboxgl/core/lib/MapEvents.js
'use strict';

exports.__esModule = true;

var _map2 = require('lodash/map');

var _map3 = _interopRequireDefault(_map2);

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Children = require('./Children');

var _Children2 = _interopRequireDefault(_Children);

var _MapEvent = require('./MapEvent');

var _MapEvent2 = _interopRequireDefault(_MapEvent);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var MapEvents = (_temp = _class = function (_React$PureComponent) {
  _inherits(MapEvents, _React$PureComponent);

  function MapEvents() {
    _classCallCheck(this, MapEvents);

    return _possibleConstructorReturn(this, _React$PureComponent.apply(this, arguments));
  }

  MapEvents.prototype.render = function render() {
    var _this2 = this;

    return _react2.default.createElement(
      _Children2.default,
      null,
      (0, _map3.default)(MapEvents.propTypes, function (_, type) {
        return _this2.props[type] ? _react2.default.createElement(_MapEvent2.default, {
          key: type,
          type: type.slice(2).toLowerCase(),
          onChange: _this2.props[type]
        }) : null;
      })
    );
  };

  return MapEvents;
}(_react2.default.PureComponent), _class.propTypes = {
  onResize: _propTypes2.default.func,
  onDblClick: _propTypes2.default.func,
  onClick: _propTypes2.default.func,
  onMouseMove: _propTypes2.default.func,
  onMoveStart: _propTypes2.default.func,
  onMove: _propTypes2.default.func,
  onMoveEnd: _propTypes2.default.func,
  onMouseOut: _propTypes2.default.func,
  onMouseDown: _propTypes2.default.func,
  onMouseUp: _propTypes2.default.func,
  onDragStart: _propTypes2.default.func,
  onDrag: _propTypes2.default.func,
  onDragEnd: _propTypes2.default.func,
  onTouchMove: _propTypes2.default.func,
  onTouchStart: _propTypes2.default.func,
  onTouchEnd: _propTypes2.default.func,
  onTouchCancel: _propTypes2.default.func,
  onZoomStart: _propTypes2.default.func,
  onZoom: _propTypes2.default.func,
  onZoomEnd: _propTypes2.default.func,
  onBoxZoomStart: _propTypes2.default.func,
  onBoxZoomEnd: _propTypes2.default.func,
  onBoxZoomCancel: _propTypes2.default.func,
  onRotateStart: _propTypes2.default.func,
  onRotate: _propTypes2.default.func,
  onRotateEnd: _propTypes2.default.func,
  onPitchStart: _propTypes2.default.func,
  onPitch: _propTypes2.default.func,
  onPitchEnd: _propTypes2.default.func,
  onWebGLContextLost: _propTypes2.default.func,
  onWebGLContextRestored: _propTypes2.default.func,
  onRemove: _propTypes2.default.func,
  onStyleData: _propTypes2.default.func,
  onSourceData: _propTypes2.default.func,
  onSourceDataLoading: _propTypes2.default.func,
  onData: _propTypes2.default.func,
  onDataLoading: _propTypes2.default.func,
  onRender: _propTypes2.default.func,
  onContextMenu: _propTypes2.default.func,
  onStyleDataLoading: _propTypes2.default.func,
  onError: _propTypes2.default.func
}, _class.contextTypes = {
  map: _propTypes2.default.object
}, _temp);
exports.default = MapEvents;
