// @react-mapboxgl/core/lib/MapGL.js
'use strict';

exports.__esModule = true;

var _keys2 = require('lodash/keys');

var _keys3 = _interopRequireDefault(_keys2);

var _pick2 = require('lodash/pick');

var _pick3 = _interopRequireDefault(_pick2);

var _extend2 = require('lodash/extend');

var _extend3 = _interopRequireDefault(_extend2);

var _isEqual2 = require('lodash/isEqual');

var _isEqual3 = _interopRequireDefault(_isEqual2);

var _class, _temp2; /**
                     * A minor clean-up of react-mapbox-gl/map.
                     */


var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _MapEvents = require('./MapEvents');

var _MapEvents2 = _interopRequireDefault(_MapEvents);

var _MapOptions = require('./MapOptions');

var _MapOptions2 = _interopRequireDefault(_MapOptions);

var _MapPosition = require('./MapPosition');

var _MapPosition2 = _interopRequireDefault(_MapPosition);

var _MapInteraction = require('./MapInteraction');

var _MapInteraction2 = _interopRequireDefault(_MapInteraction);

var _Children = require('./Children');

var _Children2 = _interopRequireDefault(_Children);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var MapGL = (_temp2 = _class = function (_React$Component) {
  _inherits(MapGL, _React$Component);

  function MapGL() {
    var _temp, _this, _ret;

    _classCallCheck(this, MapGL);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, _React$Component.call.apply(_React$Component, [this].concat(args))), _this), _this.state = {
      unsupported: false,
      map: null
    }, _this.getChildContext = function () {
      return {
        map: _this.state.map
      };
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  MapGL.prototype.shouldComponentUpdate = function shouldComponentUpdate(nextProps, nextState) {
    return !(0, _isEqual3.default)(this.props, nextProps) || !(0, _isEqual3.default)(this.state, nextState);
  };

  MapGL.prototype.componentDidMount = function componentDidMount() {
    var mapboxgl = this.context.mapboxgl;

    if (mapboxgl.supported()) {
      this.createMap();
    } else {
      this.setState({ unsupported: true });
    }
  };

  MapGL.prototype.componentWillUnmount = function componentWillUnmount() {
    var map = this.state.map;

    this.unmounted = true;
    if (map) {
      map.off();
      // NOTE: We need to defer removing the map to after all
      // children have unmounted
      setImmediate(function () {
        map.remove();
      });
    }
  };

  MapGL.prototype.componentWillReceiveProps = function componentWillReceiveProps(nextProps) {
    var map = this.state.map;

    if (map && !(0, _isEqual3.default)(this.props.containerStyle, nextProps.containerStyle)) {
      map.resize();
    }
  };

  MapGL.prototype.createMap = function createMap() {
    var _this2 = this;

    var mapboxgl = this.context.mapboxgl;

    // Build options from sub-components.

    var options = (0, _extend3.default)({ container: this.container }, _MapOptions2.default.getOptions(this.props), _MapPosition2.default.getOptions(this.props), _MapInteraction2.default.getOptions(this.props));

    // Create map.
    var map = new mapboxgl.Map(options);

    // On map load, trigger events and set state.
    map.on('load', function () {
      if (!_this2.unmounted) {
        if (_this2.props.onLoad) {
          var _props;

          (_props = _this2.props).onLoad.apply(_props, arguments);
        }
        _this2.setState({ map: map });
      }
    });

    // Optionally handle style.load.
    if (this.props.onStyleLoad) {
      map.on('style.load', this.props.onStyleLoad);
    }

    return map;
  };

  MapGL.prototype.renderUnsupported = function renderUnsupported() {
    return this.props.renderUnsupported ? this.props.renderUnsupported() : _react2.default.createElement(
      'div',
      { className: 'unsupported' },
      'Your browser does not support WebGL-based maps.'
    );
  };

  MapGL.prototype.render = function render() {
    var _this3 = this;

    var _props2 = this.props,
        containerStyle = _props2.containerStyle,
        className = _props2.className;
    var _state = this.state,
        unsupported = _state.unsupported,
        map = _state.map;

    return _react2.default.createElement(
      'div',
      {
        ref: function ref(x) {
          _this3.container = x;
        },
        className: 'react-mapbox--container ' + (className || ''),
        style: containerStyle
      },
      unsupported ? this.renderUnsupported() : map ? _react2.default.createElement(
        _Children2.default,
        null,
        _react2.default.createElement(_MapOptions2.default, (0, _pick3.default)(this.props, (0, _keys3.default)(_MapOptions2.default.propTypes))),
        _react2.default.createElement(_MapPosition2.default, (0, _pick3.default)(this.props, (0, _keys3.default)(_MapPosition2.default.propTypes))),
        _react2.default.createElement(_MapInteraction2.default, (0, _pick3.default)(this.props, (0, _keys3.default)(_MapInteraction2.default.propTypes))),
        _react2.default.createElement(_MapEvents2.default, (0, _pick3.default)(this.props, (0, _keys3.default)(_MapEvents2.default.propTypes))),
        this.props.children
      ) : null
    );
  };

  return MapGL;
}(_react2.default.Component), _class.propTypes = {
  containerStyle: _propTypes2.default.object,
  loadCSS: _propTypes2.default.bool,
  renderUnsupported: _propTypes2.default.func,
  onLoad: _propTypes2.default.func,
  onStyleLoad: _propTypes2.default.func
  // MapOptions
  // MapPosition
  // MapInteraction
  // MapEvents
}, _class.defaultProps = {
  loadCSS: true,
  containerStyle: {
    position: 'relative',
    width: '100%',
    paddingBottom: '50%'
  }
}, _class.contextTypes = {
  mapboxgl: _propTypes2.default.object
}, _class.childContextTypes = {
  map: _propTypes2.default.object
}, _temp2);
exports.default = MapGL;
